server/datastore/mysql/migrations/tables/20230719152138_ScheduledQueriesDataMigrator.go

```diff
@@ -0,0 +1,268 @@
+	// Queries can either be:
+	// 	1 - Scheduled, which can either belong to:
+	//		1.1 - The global pack:
+	// 			For each scheduled query create a single global scheduled query named `$query.name (Global Pack - $scheduled.id)`.
+	//		1.2 - Team pack:
+	//			Create a new team query with the name of `$query.name (Team Pack - $scheduled.id)`.
+	//		1.3 - A user pack (a.k.a 2017 pack):
+	// 			Do nothing.
+	//	2 - Not scheduled:
+	// 		2.1 - If the author belongs to the global team, do nothing.
+	// 		2.2 - Otherwise, for each team the author belongs to:
+	//			Create a new team query with the name of `$query.name` iff the author can run the query.
+	//
+
+	// -------------------------------------
+	// (1.1) Global pack scheduled queries
+	// -------------------------------------
+	type QueryWithScheduledFields struct {
+		fleet.Query
+		ScheduledQID       uint   `db:"scheduled_query_id"`
+		ScheduledQInterval uint   `db:"scheduled_query_interval"`
+		ScheduledQSnapshot *bool  `db:"scheduled_query_snapshot"`
+		ScheduledQRemoved  *bool  `db:"scheduled_query_removed"`
+		ScheduledQPlatform string `db:"scheduled_query_platform"`
+		ScheduledQVersion  string `db:"scheduled_query_version"`
+		PackType           string `db:"pack_type"`
+	}
+
+	selectStmt := `
+		SELECT DISTINCT q.*,
+                sq.id         AS scheduled_query_id,
+                sq.interval   AS scheduled_query_interval,
+                sq.snapshot   AS scheduled_query_snapshot,
+                sq.removed    AS scheduled_query_removed,
+                sq.platform   AS scheduled_query_platform,
+                sq.version    AS scheduled_query_version,
+				p.pack_type   AS pack_type
+		FROM queries q
+				INNER JOIN scheduled_queries sq ON q.team_id IS NULL AND q.name = sq.query_name
+				INNER JOIN packs p ON sq.pack_id = p.id
+		WHERE p.pack_type = 'global'`
+
+	insertStmt := `
+		INSERT INTO queries (
+			name,
+			description,
+			query,
+			author_id,
+			saved,
+			platform,
+			min_osquery_version,
+			schedule_interval,
+			logging_type,
+			automations_enabled 
+		) VALUES ( ?, ?, ?, ?, ?, ?, ?, ?, ?, 1 )`
+
+	rows, err := tx.Query(selectStmt)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var query QueryWithScheduledFields
+		if err := rows.Scan(&query); err != nil {
+			return err
+		}
+
+		var loggingType string
+		if query.ScheduledQSnapshot != nil && *query.ScheduledQSnapshot {
+			loggingType = "snapshot"
+		}
+		if query.ScheduledQRemoved != nil {
+			if *query.ScheduledQRemoved {
+				loggingType = "differential"
+			} else {
+				loggingType = "differential_ignore_removals"
+			}
+		}
+		args := []interface{}{
+			fmt.Sprintf("%s (Global Pack - %d)", query.Name, query.ScheduledQID),
+			query.Description,
+			query.Query,
+			query.AuthorID,
+			query.Saved,
+			query.ScheduledQPlatform,
+			query.ScheduledQVersion,
+			query.ScheduleInterval,
+			loggingType,
+		}
+		_, err = tx.Exec(insertStmt, args...)
```

Comment:
Why do we need global (scheduled) queries to be duplicated? (Shouldn't they be edited?)

After this the global queries list will have the original without schedule and the copies with schedule.

Reply:
So if we have something like this:
![image](https://github.com/fleetdm/fleet/assets/8041318/68e052c4-e83d-457e-b310-5542d07cf443)

We wanted to end-up with something like this, yes?:
![image](https://github.com/fleetdm/fleet/assets/8041318/e0c33d05-1b19-490d-950b-70f9a03f6f51)

We duplicate `Query` for each schedule referenced in the global pack (adding the schedule ID as post-fix to avoid collisions). The original is left in place in case is referenced by a 2017, but we should update it's schedule properties to match those in `Schedule 3` (we need to keep the schedule info in `Schedule 3` and the original `Query` in sync as well).

Reply:
We decided that `Query` will be left there with no schedule attributes set (those are set in `Schedule 3` in the `scheduled_queries` table).

If the user edits `Query` and sets an interval != 0 && automations_enabled = true, then the query will be executed as part of its own schedule and also as part of the 2017 pack.